### PR TITLE
Fixes #742: Removed granted read buffer tracking in http2 and tcp ada…

### LINKED ISSUE #742: Remove granted read buffer tracking in http2 and tcp adaptors
The read buffers are tracked in the http2 and tcp adaptors using a list called granted_read_buffs.

Instead of tracking these buffers, modify the router shutdown to delete all adaptor connectors and listeners before the router shuts down. 
This will make sure that the buffers are cleaned up upon the receipt of an empty PN_RAW_CONNECTION_READ event on connection close.

This assumes that deletion of connectors deletes the actual connection as well.

### PR DESCRIPTION
**Pull request by ganeshmurthy**
…ptors

### CODE REVIEW

**Review by jiridanek — COMMENTED** [review]
> Here is the test failure -
https://github.com/skupperproject/skupper-router/actions/runs/3146738179/jobs/5115896447#step:29:2330
The timeout of 400 seconds, is it set here - https://github.com/skupperproject/skupper-router/blob/main/.github/workflows/build.yaml#L167
Is that an individual timeout per test of 400 seconds ?

Yes, SkstatTest::test_yy_query_many_links had 400 seconds to finish, but it did not finish. I logged issue for this test already https://github.com/skupperproject/skupper-router/issues/637

Also, I had some thoughts in https://github.com/skupperproject/skupper-router/issues/672, if we had that, it would be easier to debug this. Currently we cannot tell if this problem is caused by https://github.com/skupperproject/skupper-router/issues/535 or if it is a completely new issue.
 
> Also, the PR is attempting to close all outstanding adaptor (tcp, http) listeners, connectors and non amqp connections before the router shuts down in test. This way, I don't have to track granted read buffers in the adaptor code. But those buffers are still leaking here - https://github.com/skupperproject/skupper-router/pull/743/checks#step:6:505
It looks like we will have to do the same thing (shutdown all connections) in the container image test that is running h2spec.

What the test does now is described in the comment https://github.com/skupperproject/skupper-router/blob/000de22ce0049c2079d462fca669cc85f0be07e9/tests/testcontainers/test_h2spec.rs#L92-L99

What you suggest is that we need to run skmanage, find out what connections are active, and terminate the connections using a management operation? Or is the trick to shutdown the nghttpd first, and then shutdown the router? How do I detect that the connection is already scrapped in the router? With a skmanage call?

Wouldn't it be simpler to teardown all connections in the router itself if it is running in a memory-debug mode?

**Comment by jiridanek** [comment]
@ganeshmurthy When I put https://github.com/skupperproject/skupper-router/pull/750 on top of this, the leak from image tests disappears for me

**Comment by ganeshmurthy** [comment]
> https://github.com/skupperproject/skupper-router/actions/runs/3159057422/jobs/5141859848#step:10:5046
> 
> ```
> 33/33 Test #57: system_tests_http1_adaptor ...........***Timeout 1199.71 sec
> ```
> 
> What just happened?

I don't think that failure has anything to do with this PR in particular but I am not sure. This PR is trying to shut down all connections just before the router shuts down, so I am hoping that the PR did not cause the error. @kgiusti, what do you think ?

File: src/adaptors/adaptor_common.h [hunk 1]
```diff
@@ -86,14 +86,11 @@ bool qd_tls_initial_setup(qd_adaptor_config_t *config,
 
 /**
  * Grants as many read qd_adaptor buffers as returned by pn_raw_connection_read_buffers_capacity() with a maximum of
- * RAW_BUFFER_BATCH(16) read buffers. Stuffs the granted adaptor buffers into the passed in granted_read_buffs list if
- * provided.
+ * RAW_BUFFER_BATCH(16) read buffers.
```
Comment (kgiusti) [inline src/adaptors/adaptor_common.h:89]: > with a maximum of RAW_BUFFER_BATCH(16) read buffers.

Really? Reading the code that doesn't seem to be the case - it will always grant read_buffers_capacity().  It is not limited to RAW_BUFFER_BATCH - it simply passes at most RAW_BUFFER_BATCH buffers in one call to pn_raw_connection_give_read_buffers. The outer loop calls pn_raw_connection_give_read_buffers() multiple times to fill the entire capacity.

Which is correct - it should fill the capacity. My point is that it appears that the comment is wrong.

File: src/router_core/connections.c [hunk 2]
```diff
@@ -273,8 +273,14 @@ void qdr_record_link_credit(qdr_core_t *core, qdr_link_t *link)
 void qdr_close_connection_CT(qdr_core_t *core, qdr_connection_t  *conn)
 {
     sys_mutex_lock(&conn->work_lock);
-    conn->closed = true;
-    conn->error  = qdr_error(QD_AMQP_COND_CONNECTION_FORCED, "Connection forced-closed by management request");
+    //
+    // In some cases, this function can be called on an already closed connection in which case
```
Comment (kgiusti) [inline src/router_core/connections.c:277]: Calling this twice seems like a bug - is this new behavior caused by this patch?  The adaptors are not calling the qdr_core_close_connection() API twice on the same qdr_connection_t, are they?  Or is the management agent attempting to re-delete a deleted connection?

**Review by kgiusti — COMMENTED** [review]
I like what this does, but I'm concerned that there's something not right if the router is calling close connection twice now.

I would also break this up into two separate PRs: one patch to the CI that does the adaptor/connector/connection cleanup and a follow up PR patch that removes the adaptor buffer tracking.  These really are two separate fixes.

File: tests/system_test.py [hunk 3]
```diff
@@ -66,8 +66,14 @@
 from skupper_router.management.client import Node
 from skupper_router.management.error import NotFoundStatus
 
+# Optional modules
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply tests/system_test.py:69]: Not sure how that got added, will remove

File: src/router_core/connections.c (hunk 2 above)
Comment (ganeshmurthy) [reply src/router_core/connections.c:277]: It is the latter and it happens in the http2 adaptor. When a connector is deleted in the http2 adaptor, it immediately (but asynchronously tries) to delete the associated connection (in http2 adaptor, one connector can have only one associated connection). While the core thread is actually trying to delete the connection, the management command comes in to tries to delete the connection as well. This happened 1 out of 10 times in my testing, so I added that bit code to check if the conn is already closed to prevent a qdr_error leak.

**Comment by ganeshmurthy** [comment]
> I like what this does, but I'm concerned that there's something not right if the router is calling close connection twice now.
> 
> I would also break this up into two separate PRs: one patch to the CI that does the adaptor/connector/connection cleanup and a follow up PR patch that removes the adaptor buffer tracking. These really are two separate fixes.

The reason, I had all this in the same patch is because these are related. Deleting the connectors/connections is what is allowing me to take out the tracking of the granted_read_buffers from the adaptors. If I do a PR with the management change separately and another PR which just removes the granted_read_buffs, I am sure someone might wonder looking at the second PR why suddenly we are able to remove the granted_read_buffs.

File: tests/system_test.py [hunk 4]
```diff
@@ -594,6 +613,22 @@ def teardown(self):
             self._management = None
 
         teardown_exc = None
+
+        try:
+            qd_manager = QdManager(address=self.addresses[0])
```
[the remark this answers is not in the document]
Comment (kgiusti) [reply tests/system_test.py:618]: Related #753

File: tests/system_test.py [hunk 5]
```diff
@@ -594,6 +613,22 @@ def teardown(self):
...
+            # Delete all adaptor connectors and listeners before shutting down the router
+            long_types = [HTTP_LISTENER_TYPE, HTTP_CONNECTOR_TYPE, TCP_LISTENER_TYPE, TCP_CONNECTOR_TYPE]
+            for long_type in long_types:
+                qd_manager.delete_all_entities(long_type)
+            retry_assertion(self.delete_adaptor_connections(qd_manager))
+        except Exception as e:
+            # We tried to delete some stuff but ran into issues. There could
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/system_test.py:625]: +1, also, catching `Exception` will catch SyntaxError, etc... Best to have explicit list, esp. because it is hard to improve this later on

File: src/router_core/connections.c (hunk 2 above)
Comment (kgiusti) [reply src/router_core/connections.c:277]: Ah! Ok, so if I understand correctly:

1. The CI change adds a management call to delete the _connection_ - which is cool.
2. At the same moment the adaptor is servicing a previous management request to delete the _connector_, which then correctly calls qdr_core_close_connection()

So at the core we have two connection delete paths - one via the management agent and one via the action queued by the adaptor thread - both are valid and can potentially happen simultaneously "in the field".

Given that I agree the check for conn->closed is the correct move.

But - and it seems I always have a big but, right? - I wouldn't activate the connection if conn->closed was already true. That will cause the connection process to call the close handler twice which it probably shouldn't do.

See next comment.

File: src/router_core/connections.c [hunk 6]
```diff
@@ -273,8 +273,14 @@ void qdr_record_link_credit(qdr_core_t *core, qdr_link_t *link)
...
+    // we don't want to leak a qdr_error_t object
+    //
+    if (!conn->closed) {
+        conn->closed = true;
```
Comment (kgiusti) [inline src/router_core/connections.c:281]: In here I'd try setting a local flag to call the activate once the lock is dropped.  That would ensure that the close handler is not invoked twice.

Try that.  Does it break anything?

File: src/adaptors/adaptor_common.h (hunk 1 above)
Comment (ganeshmurthy) [reply src/adaptors/adaptor_common.h:89]: I fixed the comment.

File: src/router_core/connections.c (hunk 6 above)
Comment (ganeshmurthy) [reply src/router_core/connections.c:281]: I used a local flag to call activate outside the lock and it does the job, does not break anything. I am running the http2 and http2_tls tests in a loop with -j4 and will run it for some more time to see if it catches anything

File: tests/system_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply tests/system_test.py:625]: Sometimes a proton ConnectionException is thrown which proton does not seem to expose ?

File: tests/system_test.py [hunk 7]
```diff
@@ -1453,6 +1488,14 @@ def get_log(self, limit=None):
...
         return json.loads(self(cmd))
 
+    def delete_all_entities(self, long_entity_type):
+        # First, query the router to get all entities of entity type
+        results = self.query(long_entity_type)
+        for result in results:
+            entity_name = result.get('name')
+            if entity_name:
+                self.delete(long_entity_type, name=entity_name)
+
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply tests/system_test.py:1498]: Since the delete command is synchronous, the delete is successful (meaning the entity is deleted) by the time the response to the delete command reaches qdmanage. So, does adding an additional retry loop buy us anything since we already know that the entity has already been deleted ?
If the delete command is going to start a core thread action, we never know when that is going to complete ?

File: tests/system_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/system_test.py:625]: It should be exposed

https://github.com/apache/qpid-proton/blob/a4375a8351c3435bd5025bb2481307efe97dc99e/python/proton/__init__.py#L57

so

```
from proton import ConnectionException
```

?

**Comment by kgiusti** [comment]
> > https://github.com/skupperproject/skupper-router/actions/runs/3159057422/jobs/5141859848#step:10:5046
> > ```
> > 33/33 Test #57: system_tests_http1_adaptor ...........***Timeout 1199.71 sec
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What just happened?
> 
> I don't think that failure has anything to do with this PR in particular but I am not sure. This PR is trying to shut down all connections just before the router shuts down, so I am hoping that the PR did not cause the error. @kgiusti, what do you think ?

Oh great... yet another buffer corruption issue!

From the server log:

2022-09-30 13:26:40.230373 +0000 HTTP_ADAPTOR (trace) [C1] HTTP request/response codec done. **Octets read: 393570 written: 210** (../src/adaptors/http1/http1_server.c:1176)

Yep, those numbers look good.  Now what did the client see?  From the client log:

2022-09-30 13:31:40.304957 +0000 HTTP_ADAPTOR (trace) [C66] HTTP request msg-id=18 cancelled. **Octets read: 210 written: 195584** (../src/adaptors/http1/http1_client.c:965)

TL;DR test client is hanging since it hasn't received the full response although the server sent it.

I'll go re-open the old Issue... Oh look! It's about Beer O'clock on a Friday!  So long, folks!!

File: src/router_core/connections.c (hunk 2 above)
Comment (ganeshmurthy) [reply src/router_core/connections.c:278]: Used a local flag (outside of lock) to avoid activation and it worked !

File: tests/system_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply tests/system_test.py:1499]: Also, there  is a loop that deletes all connections as well.

File: tests/system_test.py [hunk 8]
```diff
@@ -578,6 +584,15 @@ def __enter__(self):
     def __exit__(self, *_):
         self.teardown()
 
+    @property
+    def qd_manager(self):
+        if not self._qd_manager:
+            try:
+                self._qd_manager = QdManager(address=self.addresses[0])
+            except Exception as e:
+                return None
```
Comment (kgiusti) [inline tests/system_test.py:593]: Probably don't want to silently discard the exception - it will be very hard to figure out what happened if the QdManager fails in GH actions.

File: tests/system_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kgiusti) [reply tests/system_test.py]: If I understand correctly the problem is that qdmanager is used to do the operation. Qdmanager isn't using proton directly - it spawns a process that runs skmanage as a shell process then simply raises Exception should the process return an error code.

So we lose any access to the proton exception details.

File: tests/system_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kgiusti) [reply tests/system_test.py]: There are multiple ways of de-furring this particular feline. I'll leave it up to the judgment of whomever is brave enough to take on #753.

File: tests/system_test.py (hunk 8 above)
Comment (ganeshmurthy) [reply tests/system_test.py:593]: > Probably don't want to silently discard the exception - it will be very hard to figure out what happened if the QdManager fails in GH actions.

You mean, it will be hard to figure out why the QdManager constructor fails in GHA ? The only one reason it could fail is if there is no self.addresses, right ? I will add a print(e) before returning None

Comment (kgiusti) [reply tests/system_test.py:593]: Is not having a self.addresses[0] a bug, or is it OK as long as the caller checks for a None?  I think that's my point of confusion: I don't understand why the exception is being ignored. If the API allows a None return code (not a bug), then I would simply comment the code to explain that is why the exception is discarded.

File: tests/system_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply tests/system_test.py]: QdManager does indeed wrap all exceptions into an Exception object but the exception details are not lost. QdManager calls ```raise Exception("%s %s" % rc)``` upon failure which the following error message (for example a ConnectionException) prints - 
```ConnectionException: Connection amqp://0.0.0.0:23860 disconnected: Condition('proton.pythonio', 'Connection refused to all addresses')``` - prints on a proton ConnectionException

**Review by kgiusti — APPROVED** [review]

File: tests/system_test.py (hunk 8 above)
Comment (ganeshmurthy) [reply tests/system_test.py:593]: As it stands now, not all routers have a valid self.addresses[0], so it is ok to return a None. I will add a comment clarifying that. But the person who fixes https://github.com/skupperproject/skupper-router/issues/753 will make not having a self.addresses[0] illegal and require every router to have a management port on self.addresses[0]
